File: frontend/src/types/product.ts
export type Product = {
  id: string
  title: string
  description: string
  category: string
  tags: string[]
  price: number | null
  sku: string | null
  created_at: string
}

export type SimilarItem = {
  product_id: string
  score: number
}

export type ContentRecommendationResponse = {
  source_product_id: string
  items: SimilarItem[]
}

export type CollaborativeItem = {
  product_id: string
  estimated_rating: number
}

export type CollaborativeRecommendationResponse = {
  user_id: string
  items: CollaborativeItem[]
}

export type HybridItem = {
  product_id: string
  hybrid_score: number
  collaborative_norm: number
  content_similarity: number
  estimated_rating: number
}

export type HybridRecommendationResponse = {
  user_id: string
  weight_collaborative: number
  weight_content: number
  items: HybridItem[]
}

export type PrecisionAtKReport = {
  k: number
  users_evaluated: number
  mean_precision_at_k: number
  detail: string
}

File: backend/app/schemas/product.py
from datetime import datetime

from pydantic import BaseModel, Field


class ProductCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=500)
    description: str = Field(default="", max_length=20000)
    category: str = Field(default="", max_length=200)
    tags: list[str] = Field(default_factory=list)
    price: float | None = Field(default=None, ge=0)
    sku: str | None = Field(default=None, max_length=120)


class ProductOut(BaseModel):
    id: str
    title: str
    description: str
    category: str
    tags: list[str]
    price: float | None
    sku: str | None
    created_at: datetime
